### PR TITLE
Tolerations for Shoots and Projects

### LINKED ISSUE #401: Configurable list for namespaces which may use protected seeds
The Seed spec allows to configure `.spec.protected={true,false}` which means that only Shoots that have been deployed to the `garden` namespace may use this respective Seed cluster.

We should make this list of namespaces configurable, either globally (in the configuration of the SeedManager admission plugin) or per Seed (maybe in the resource directly, maybe also in the admission plugin's configuration).

### LINKED ISSUE #2193: Project Tolerations to support limiting access to Seed usage
**What would you like to be added**:
Today, the `Seed` can have several taints specified in the `.spec.taints[]` list. One of them is `seed.gardener.cloud/protected`. If set, only shoots in the `garden` namespace are allowed to specify the seed in the `.spec.seedName`. However, the `garden` namespace is hard-coded (see #401) which is not good. Also, there is a new use case where specific seeds shall be restricted to specific (multiple) projects only.

Let's adapt our `Project` resource with a new `.spec.tolerations[]` field:

```yaml
---
apiVersion: core.gardener.cloud/v1beta1
kind: Project
metadata:
  name: my-project
spec:
  ...
  tolerations:
  - key: <some-key>
```

If all taints of a `Seed` are tolerated by the `Project` then the `Shoot` is allowed to use it. Today, we also have other taints (`invisible`, `disable-dns`, `excess-capacity`). These are more settings of the seed and don't fit to the taint/toleration concept. Hence, we should move them out of the `.spec.taints[]` to a new `.spec.settings[]` field.

Now a `Seed` can have the `<seed-name>.seed.gardener.cloud/protected` taint. If the project has the `<seed-name>.seed.gardener.cloud/protected` toleration then a `Shoot` is allowed to be scheduled on it.

In order to prevent that end-users (which are having `update`,`patch` privileges) add arbitrary `tolerations` we should use a custom RBAC verb. Only with it it would be allowed to change the `.spec.tolerations` field:

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: gardener.cloud:project-tolerations
rules:
- apiGroups:
  - core.gardener.cloud
  resources:
  - projects
  verbs:
  - tolerations
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: gardener.cloud:project-tolerations
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: gardener.cloud:project-tolerations
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: gardener-admin
```

The scheduler and our admission should be enhanced to respect these new taints/tolerations concepts.

PS: In a future iteration we could think about introducing a new namespaced `Seed` resource, but this is out of scope of this issue.

**Why is this needed**:
Exclude projects from using various seeds to make them exclusive to some of them.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability
/area documentation
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR adds support for tolerations for `Shoot`s, similar to tolerations for `Pod`s.

Also, it supports specifying a whitelist + defaults for on `Project` level. Only users with the `modify-spec-tolerations-whitelist` custom verb are allowed to change the whitelist. This is validated by a new `CustomVerbAuthorizer` admission plugin.

The new `ShootTolerationRestriction` admission plugin validates that only whitelisted tolerations might be used by shoots. It also defaults them according to the default tolerations configured in the `Project`. It also allows globally configured defaults + whitelist with its own admission plugin configuration.

I added some documentation about all the existing admission plugins along the way.

**Which issue(s) this PR fixes**:
Fixes #2193
Fixes #401

**Release note**:

```noteworthy user
`Shoot` resources do now support specifying tolerations for taints on seeds in the `.spec.tolerations` field. Only tolerations that were whitelisted by the corresponding `Project`'s `.spec.tolerations.whitelist` field, or by the global configuration (controlled by Gardener administrators) are allowed to be used. Please read more about it in [this document](https://github.com/gardener/gardener/blob/master/docs/usage/tolerations.md).
```
```improvement operator
A new [document](https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver_admission_plugins.md) providing an overview of the existing in-tree admission plugins in Gardener has been added.
```

### CODE REVIEW

File: docs/concepts/apiserver_admission_plugins.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Admission Plugins
+
+Similar to the kube-apiserver, the gardener-apiserver comes with a few in-tree managed admission plugins.
+If you want to get an overview of the what and why of admission plugins then [this document](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) might be a good start.
+
+This document enlists all existing admission plugins with a short explanation of what it is responsible for.
```
Comment (timuthy) [inline docs/concepts/apiserver_admission_plugins.md]:
```suggestion
This document lists all existing admission plugins with a short explanation of what it is responsible for.
```

File: pkg/apis/core/helper/helpers.go [hunk 2]
```diff
@@ -133,6 +134,41 @@ func TaintsHave(taints []core.SeedTaint, key string) bool {
...
+		if taint.Key == core.DeprecatedSeedTaintDisableDNS || taint.Key == core.DeprecatedSeedTaintInvisible || taint.Key == core.DeprecatedSeedTaintDisableCapacityReservation {
+			continue
+		}
+		relevantTaints = append(relevantTaints, taint)
+	}
+
+	if len(relevantTaints) == 0 {
+		return true
+	}
+	if len(tolerations) == 0 {
```
Comment (timuthy) [inline pkg/apis/core/helper/helpers.go]:
```suggestion
	if len(relevantTaints) > len(tolerations) {
```

File: pkg/apis/core/helper/helpers_test.go [hunk 3]
```diff
@@ -93,6 +93,96 @@ var _ = Describe("helper", func() {
...
+			[]core.SeedTaint{
+				{Key: "foo"},
+				{Key: "bar", Value: pointer.StringPtr("baz")},
+			},
+			[]core.Toleration{
+				{Key: "foo", Value: pointer.StringPtr("bar")},
+				{Key: "bar", Value: pointer.StringPtr("foo")},
+			},
+			false,
+		),
```
Comment (timuthy) [inline pkg/apis/core/helper/helpers_test.go:183]:
```suggestion
		Entry("taints > tolerations",
			[]core.SeedTaint{
				{Key: "foo"},
				{Key: "bar", Value: pointer.StringPtr("baz")},
			},
			[]core.Toleration{
				{Key: "bar", Value: pointer.StringPtr("baz")},
			},
			false,
		),
	),
```

File: pkg/apis/core/v1alpha1/defaults.go [hunk 4]
```diff
@@ -181,6 +191,10 @@ func SetDefaults_Shoot(obj *Shoot) {
 		p := ShootPurposeEvaluation
 		obj.Spec.Purpose = &p
 	}
+
+	if obj.Namespace == v1beta1constants.GardenNamespace {
+		addTolerations(&obj.Spec.Tolerations, Toleration{Key: SeedTaintProtected})
```
Comment (timuthy) [inline pkg/apis/core/v1alpha1/defaults.go:199]: Can you add a comment which explains the purpose of this action? I think it is or will not be obvious in the future.

File: pkg/apis/core/v1beta1/helper/helper.go [hunk 5]
```diff
@@ -211,6 +213,41 @@ func TaintsHave(taints []gardencorev1beta1.SeedTaint, key string) bool {
...
+		if taint.Key == gardencorev1beta1.DeprecatedSeedTaintDisableDNS || taint.Key == gardencorev1beta1.DeprecatedSeedTaintInvisible || taint.Key == gardencorev1beta1.DeprecatedSeedTaintDisableCapacityReservation {
+			continue
+		}
+		relevantTaints = append(relevantTaints, taint)
+	}
+
+	if len(relevantTaints) == 0 {
+		return true
+	}
+	if len(tolerations) == 0 {
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/helper/helper.go]:
```suggestion
	if len(relevantTaints) > len(tolerations) {
```

File: plugin/pkg/global/customverbauthorizer/admission.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		if !ok {
+			return apierrors.NewBadRequest("could not convert old resource into Project object")
+		}
+	}
+
+	if apiequality.Semantic.DeepEqual(oldObj.Spec.Tolerations, obj.Spec.Tolerations) {
+		return nil
+	}
+
+	return c.authorize(ctx, a, "modify-spec-tolerations-whitelist", "modify .spec.tolerations.whitelist")
```
Comment (timuthy) [inline plugin/pkg/global/customverbauthorizer/admission.go]: `modify-spec-tolerations-whitelist`, can you make it a `const`?

File: plugin/pkg/shoot/tolerationrestriction/admission.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	return nil
+}
+
+func (t *TolerationRestriction) admitShoot(shoot *core.Shoot) error {
+	project, err := admissionutils.GetProject(shoot.Namespace, t.projectLister)
+	if err != nil {
+		return apierrors.NewBadRequest(fmt.Sprintf("could not find referenced project: %+v", err.Error()))
+	}
+
+	// do not change shoot tolerations if they specify a key already
```
Comment (timuthy) [inline plugin/pkg/shoot/tolerationrestriction/admission.go]: Rather move comment to line 173?

File: pkg/apis/core/helper/helpers.go [hunk 8]
```diff
@@ -133,6 +134,41 @@ func TaintsHave(taints []core.SeedTaint, key string) bool {
...
+	taintKeyValues := sets.NewString()
+	for _, taint := range relevantTaints {
+		taintKeyValues.Insert(utils.IDForKeyWithOptionalValue(taint.Key, taint.Value))
+	}
+
+	for _, toleration := range tolerations {
+		if taintKeyValues.Has(toleration.Key) {
+			continue
+		}
+		if !taintKeyValues.Has(utils.IDForKeyWithOptionalValue(toleration.Key, toleration.Value)) {
```
Comment (timuthy) [inline pkg/apis/core/helper/helpers.go]: This means that all taints must contain all tolerations but it should be the other way around, shouldn't it? At the moment, a seed is rejected if the shoot has more tolerations than the seed has taints.

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice PR! I left a few comments.

File: pkg/apis/core/helper/helpers.go (hunk 8 above)
Comment (rfranzke) [reply pkg/apis/core/helper/helpers.go]: Very nice catch, thanks for finding it, indeed. I've inverted the logic as you outlined and added another test case for it.

**Review by timuthy — COMMENTED** [review]
I tested your changes again and it works very well! One restriction we should keep in mind is that because we only write the default tolerations during the creation, a `kubectl apply` with new tolerations will delete the ones we defaulted during creation.

For example:
1. Add defaults to project
2. `kubectl apply shoot.yaml` w/o tolerations
3. Ensure that default tolerations were added to shoot resource
4. Edit `shoot.yaml` and add a new (white listed) toleration
5. `kubectl apply shoot.yaml`
6. Observe that the default toleration is gone.

Unfortunately, `patchStrategy:"merge" patchMergeKey:"key"` struct tags don't help either since `kubectl apply` doesn't use the [strategic merge patch](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch). Don't consider it a blocker but I wanted to hear your opinion or maybe you have an idea how to circumvent it.

**Comment by rfranzke** [comment]
Thanks for testing! First of all, we should probably add the patch strategy anyways, right?
As for the situation you explained: I would think this expected behaviour, or? If you specify tolerations yourself (like you did in step 4) then those are being considered the relevant ones. The defaults are only added if you don't specify any toleration at all.

**Comment by timuthy** [comment]
Thanks for your thoughts. As discussed, let's handle it the way you proposed it 👍

**Review by timuthy — DISMISSED** [review]
/lgtm

**Review by timuthy — APPROVED** [review]
/lgtm

**Pull request merged**